File: problem-solving/baekjoon/1182_부분수열의_합.go
// https://www.acmicpc.net/problem/1182
package main

import (
	"bufio"
	"fmt"
	"os"
)

var (
	count int
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	var n, s int
	fmt.Fscanln(reader, &n, &s)
	var seq = make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Fscanf(reader, "%d ", &seq[i])
	}
	computeSubsequenceCount(0, 0, s, seq)
	fmt.Fprintln(writer, count)
}

func computeSubsequenceCount(index, sum, aim int, seq []int) {
	if index > len(seq)-1 {
		return
	}
	sum += seq[index]
	if sum == aim {
		count++
	}
	computeSubsequenceCount(index+1, sum, aim, seq)            // 현재 인덱스 값을 더하지 않는 경우
	computeSubsequenceCount(index+1, sum-seq[index], aim, seq) // 현재 인덱스 값을 더하는 경우
}

File: problem-solving/baekjoon/1182_부분수열의_합.py
# https://www.acmicpc.net/problem/1182
import sys

count = 0
seq = []

def compute_subsequence_count(index, sum, aim):
    global count
    if index > len(seq)-1:
        return
    sum += seq[index]
    if sum == aim:
        count += 1
    compute_subsequence_count(index+1, sum, aim)
    compute_subsequence_count(index+1, sum-seq[index], aim)

if __name__ == "__main__":
    n, s = list(map(int, sys.stdin.readline().split()))
    seq = list(map(int, sys.stdin.readline().split()))
    compute_subsequence_count(0, 0, s)
    print(count)
